third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/PingPong.cpp

```diff
@@ -73,191 +250,244 @@ void getNestedFor(Region *partition,
-  DenseMap<unsigned, SmallVector<Operation *>> loopDepthPartition1;
-  unsigned partitionId = 0;
-  for (Region *region : wsOp.getPartitionRegions()) {
-    LDBG("-- region " << region->getNumArguments());
-    // Assume default for producer.
-    if (partitionId == 0) {
-      getNestedFor(region, loopDepthPartition0);
-    } else if (partitionId == 1) {
-      getNestedFor(region, loopDepthPartition1);
+    // Find the innermost loop (maximum depth)
+    auto &loopDepth = partitionLoopDepths[iter];
+    auto maxIt = std::max_element(
+        loopDepth.begin(), loopDepth.end(),
+        [](const auto &a, const auto &b) { return a.first < b.first; });
+    if (maxIt->second.empty())
+      continue;
+
+    auto innermostForOp = dyn_cast<scf::ForOp>(maxIt->second[0]);
+    if (!innermostForOp)
+      continue;
+
+    LDBG("Searching for exp ops in innermost loop at depth " << maxIt->first);
+    // Walk through the innermost loop to find exp operations
+    Operation *expOp = nullptr;
+    Operation *expStartOp = nullptr; // exp critical region start
+    Operation *expEndOp = nullptr;   // exp critical region end
+    SmallVector<Operation *> expOps;
+    for (auto &op : innermostForOp.getBody()->without_terminator()) {
+      // Check for exp or exp2 operations
+      if (isa<math::Exp2Op, math::ExpOp>(op)) {
+        auto tensorTy = dyn_cast<RankedTensorType>(op.getOperand(0).getType());
+        if (tensorTy && tensorTy.getRank() > 1)
+          expOps.push_back(&op);
+      }
     }
-    ++partitionId;
-  }
 
-  // Verify a single ForOp at depth of 0, i.e a single outer ForOp.
-  LDBG("Found loops: " << loopDepthPartition0.size());
-  if (loopDepthPartition0.empty() || loopDepthPartition0[0].size() != 1)
-    return;
-  // Have a ForOp at depth of 1.
-  bool hasPersistent = loopDepthPartition0.find(1) != loopDepthPartition0.end();
-
-  // Find the region of gemms via [startOfGemm, endOfGemm)
-  SmallVector<Operation *> starts, ends;
-  for (unsigned iter = 0; iter < 2; ++iter) {
-    // Check partition 0 first, then partition 1.
-    // Find the innermost ForOp (i.e depth of 1 for persistent).
-    Operation *op = iter == 0 ? loopDepthPartition0[hasPersistent ? 1 : 0][0]
-                              : loopDepthPartition1[hasPersistent ? 1 : 0][0];
-    auto forOp = dyn_cast<scf::ForOp>(op);
-    // A simple heuristic for now:
-    //   Mark the start of a gemm section when hitting a DotLike op.
-    //   Mark the end of a gemm section once hitting an expensive non-dot
-    //   computation op.
-    Operation *startOfGemm = nullptr;
-    Operation *endOfGemm = nullptr;
-    for (auto &op : forOp.getBody()->without_terminator()) {
-      if (startOfGemm && endOfGemm)
+    if (expOps.empty())
+      continue;
+
+    expOp = expOps.front();
+    expStartOp = expOps.front();
+
+    // Detect the next barrier arrive operation after tmem_store
+    Operation *currentOp = expOps.back();
+    bool foundTmemStore = false;
```

Comment:
Seems like a dependency on TMEM breaks any chance of Hopper working. It would be good to partition these in some way.